include/range/v3/view/shared.hpp

```diff
@@ -91,49 +85,49 @@ namespace ranges
             struct shared_fn : pipeable<shared_fn>
             {
             public:
-                template<typename T,
-                    CONCEPT_REQUIRES_(Range<T>())>
-                shared_view<T> operator()(std::shared_ptr<T> pt) const
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(Range<Rng>())>
+                shared_view<Rng> operator()(std::shared_ptr<Rng> pt) const
                 {
                     return {std::move(pt)};
                 }
 
 #ifndef RANGES_DOXYGEN_INVOKED
-                template<typename T,
-                    CONCEPT_REQUIRES_(!Range<T>())>
-                void operator()(std::shared_ptr<T>) const
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(!Range<Rng>())>
+                void operator()(std::shared_ptr<Rng>) const
                 {
-                    CONCEPT_ASSERT_MSG(Range<T>(),
+                    CONCEPT_ASSERT_MSG(Range<Rng>(),
                         "The object on which view::shared operates must be a "
                         "model of the Range concept.");
                 }
 #endif
 
-                template<typename T,
-                    CONCEPT_REQUIRES_(Range<T>()),
-                    typename std::enable_if<std::is_rvalue_reference<T&&>::value, bool>::type = 0>
-                shared_view<typename std::remove_reference<T>::type> operator()(T && t) const
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(Range<Rng>()
+                                      && !View<Rng>()
+                                      && std::is_rvalue_reference<Rng&&>::value)>
+                shared_view<typename std::remove_reference<Rng>::type> operator()(Rng && t) const
```

Comment:
`std::is_rvalue_reference_v<Rng&&>` implies `std::is_same_v<Rng, meta::_t<std::remove_reference<Rng>>>`, so `typename std::remove_reference<Rng>::type` here (and on 112) can simply be `Rng`.